src/charm.py

```diff
@@ -0,0 +1,50 @@
+#!/usr/bin/env python3
+# Copyright 2024 dima.tisnek@canonical.com
+# See LICENSE file for licensing details.
+"""Charmed Gubernator."""
+import socket
+
+import ops
+from charms.traefik_k8s.v2.ingress import IngressPerAppRequirer
+
+
+def kubernetes_service_dns_name():
+    """Return the service DNS name for this application.
+
+    This hostname is resolved to a set of ip addresses within the kubernetes cluster.
+    Gubernator is designed so that any node can be called with any request, and it
+    will handle the synchronisation internally.
+    """
+    pod_hostname = socket.getfqdn()
+    _unit, service_dns_name = pod_hostname.split(".", 1)
+    # FIXME: not sure about this, it could be customised by k8s admin
+    assert service_dns_name.endswith(".local")
+    return service_dns_name
+
+
+class HexanatorCharm(ops.CharmBase):
+    """Charm the service."""
+
+    def __init__(self, framework: ops.Framework):
+        super().__init__(framework)
+        self.ingress = IngressPerAppRequirer(self, port=80, strip_prefix=True)
+        self.framework.observe(self.on["gubernator"].pebble_ready, self._on_pebble_ready)
+        self.framework.observe(self.on["rate-limit"].relation_created, self._on_relation)
+
+    def _on_pebble_ready(self, event: ops.PebbleReadyEvent):
+        """Kick off Pebble services.
+
+        The `gubernator` service is configured and enabled in the `rockcraft.yaml` file.
+        Pebble starts with `--on-hold` in the workload container, release it.
+        """
+        event.workload.replan()
```

Comment:
There is a very slim chance of this failing because Pebble (or the workload container) has gone away between the ready event being triggered and here. If that happened, the unit would go into error state, and the problem would probably solve itself (k8s recreation or whatever) and when the event was retried, or a new pebble-ready was received, all would be ok.

I feel like that's fine. However, charmers don't like going into error state, so here would catch the Pebble error (probably ConnectionError) and defer. Up to you what you do :)